kaiax/supply/impl/execution.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2024 The Kaia Authors
+// This file is part of the Kaia library.
+//
+// The Kaia library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Kaia library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package supply
+
+import (
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/common"
+)
+
+// PostInsertBlock will try to advance the supplyCheckpoint by one block.
+// If the new block is right after the last checkpoint, it will advance the supplyCheckpoint.
+// Otherwise, it will leave the work to the catchup thread.
+func (s *SupplyModule) PostInsertBlock(block *types.Block) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	newNum := block.NumberU64()
+	if s.lastAccNum+1 == newNum && s.lastAccReward != nil {
+		newCheckpoint, err := s.accumulateRewards(s.lastAccNum, newNum, s.lastAccReward, true)
+		if err != nil {
+			return err
+		}
+		s.lastAccNum = newNum
+		s.lastAccReward = newCheckpoint
+	}
+	return nil
+}
+
+func (s *SupplyModule) RewindTo(newBlock *types.Block) {
+	// Soft reset to the nearest checkpoint interval less than the new block number,
+	// so that the next accumulation will start below the rewound block.
+	newLastNum := nearestCheckpointNumber(newBlock.NumberU64())
+	WriteLastAccRewardNumber(s.ChainKv, newLastNum)
+}
+
+func (s *SupplyModule) RewindDelete(hash common.Hash, num uint64) {
+	if num%checkpointInterval != 0 {
+		DeleteAccReward(s.ChainKv, num)
```

Comment:
```suggestion
		DeleteAccReward(s.ChainKv, num)
		WriteLastAccRewardNumber(s.ChainKv, num)
```

Reply:
LastAccRewardNumber should be overwritten in RewindTo which is called by updateFn.

In setHeadBeyondRoot (https://github.com/kaiachain/kaia/blob/9ec35a0050193b17c5f2dfd0f2acdc070779ff01/blockchain/blockchain.go#L663-L675), it's either `updateFn` or `updateFn+delFun`.